[SYCL] Implement specialization constants.

kbobrovs opened the pull request:
Implementation is done according to:

- the language-level spec draft from Codeplay (to be substituted by a new one soon, though, but a lot of implementation will remain):
https://github.com/codeplaysoftware/standards-proposals/blob/master/spec-constant/index.md

- SPIRV:
https://www.khronos.org/registry/spir-v/specs/1.2/SPIRV.html#_a_id_specializationsection_a_specialization

- OpenCL API:
https://www.khronos.org/registry/OpenCL/sdk/2.2/docs/man/html/clSetProgramSpecializationConstant.html

Known deficiencies:
- the SPIRV interop part is not implemented yet
- only primitive types are supported
- specific spec constant setting has global effect - i.e. all programs which depend on it will use the latest set value; that's might be not what the specification proposal meant; will be fixed either with this proposed spec or with the upcoming new one

The first three commits are not really SYCL specific and separate PRs will be created. Those are also planned to go to llvm upstream.
property I/O: #1357
table library: #1358
file-table-tform: #1359

A lot of complexity stems from the fact that SPIRV identifies specialization constants by integer ID. It is impossible to assign integer IDs at translation unit compilation time because of separate compilation and potential ID conflict between TUs. So, clang instead generates special intrinsic aka get_spec_constant_value_by_name("spec constant name generated from its type ID into int header") .

Then after linkage, the sycl-post-link tool
(1) runs an LLVMIR pass which assigns integer ID to each spec constant and replaces the get_spec_constant_value_by_name intrinsic with a standard SPIRV intrinsic to get spec constant value with integer ID
(2) saves a map "name"->"int ID" into a file which is later inserted by the clang offload wrapper into the device binary descriptor created for device code (SPIRV module).

At runtime, SYCL runtime receives a request to set a spec constant with given name to some value , uses the "name"->ID map to resolve it to an integer and call piProgramSetSpecializationConstant with the corresponding int ID.

To store the "name"->ID map a "property set" library is developed, which can store arbitrary sets of properties. ClangOffloadWrapper is able to parse them and insert into the wrapper object. Later entries, manifest, build options, etc (as well as future information accompanying the device code) can move to this mechanism instead of having separate fields in the binary descriptor.

Another complexity is clang driver's inability to conveniently model multiple dynamic inputs/outputs in the action graph. The file-table-tform tool fills the gap.
See also comments in the c9a794c.

sergey-semenov commented on sycl/source/detail/program_manager/program_manager.cpp:
```diff
@@ -472,6 +484,9 @@ string_class ProgramManager::getProgramBuildLog(const RT::PiProgram &Program,
   return Log;
 }
 
+// TODO device libraries may use scpecialization constants, manifest files, etc.
```
```suggestion
// TODO device libraries may use specialization constants, manifest files, etc.
```

sergey-semenov commented on sycl/source/detail/program_manager/program_manager.hpp:
```diff
@@ -134,10 +182,33 @@ class ProgramManager {
+  // The base approach is that there is a global name->id spec constant map for
+  // each OS module. This map is populated from the name->id pairs found in
+  // device binary image upon registration at startup for this OS module.
+  // If multiple images beloning to the same OS module have a specialization
```
```suggestion
  // If multiple images belonging to the same OS module have a specialization
```

sergey-semenov commented on sycl/source/detail/program_manager/program_manager.cpp:
```diff
@@ -844,39 +885,20 @@ void ProgramManager::addImages(pi_device_binaries DeviceBinary) {
+      Imgs.reset(new std::vector<RTDeviceBinaryImageUPtr>());
+      Imgs->push_back(std::move(Img));
+    } else
+      Imgs->push_back(std::move(Img));
```
I suggest removing the else branch and moving the `push_back` line outside of `if`.

kbobrovs replied:
OK

sergey-semenov commented on sycl/source/detail/program_manager/program_manager.cpp:
```diff
@@ -959,14 +954,143 @@ void ProgramManager::dumpImage(const DeviceImage &Img, KernelSetId KSId) const {
+
+// Convenience class to iteratate via specialization constant names in the
+// binary image given in the constructor. Also implements the C string iterator
+// interface requred by some program manager APIs.
```
```suggestion
// interface required by some program manager APIs.
```

sergey-semenov commented on sycl/source/detail/program_manager/program_manager.cpp:
```diff
@@ -959,14 +954,143 @@ void ProgramManager::dumpImage(const DeviceImage &Img, KernelSetId KSId) const {
+  Bin = nullptr;
+}
+
+// Convenience class to iteratate via specialization constant names in the
```
```suggestion
// Convenience class to iterate over specialization constant names in the
```

kbobrovs commented on llvm/tools/sycl-post-link/SpecConstants.cpp:
```diff
@@ -0,0 +1,287 @@
+      }
+      DelGlob->replaceAllUsesWith(ConstantPointerNull::get(DelGlob->getType()));
+      DelGlob->removeFromParent();
+      DelGlob->deleteValue();
```
> Review for post-link part.
> Waiting for tests for new functionality.

There are some already - a lot of driver tests are fixed to reflect new features. I will write 1 more - for the IR checking, assuming this is your main concern here.

kbobrovs replied:
> Basically lines 257-258 and 253-254 are doing the same thing - they are removing llvm Value from the module.
> I thing we could merge these two places into single loop by adding `DelGlob` to `DelInsts` and make `DelInsts` to be array of values.

OK

kbobrovs replied:
> asically lines 257-258 and 253-254 are doing the same thing - they are removing llvm Value from the module.
I thing we could merge these two places into single loop by adding DelGlob to DelInsts and make DelInsts to be array of values.

Value does not have `removeFromParent()` method. So leaving as is.

kbobrovs commented on llvm/tools/sycl-post-link/SpecConstants.cpp:
```diff
@@ -0,0 +1,287 @@
+      }
+      assert(!I->mayHaveSideEffects() && "unexpected spec const IR pattern 1");
+    }
+    assert(!L && "load not met after the store");
```
> Why we are not doing this in the release?

good point

> BTW, I tried, but it's hard to understand what the code on lines 82-86 does, is it possible to elaborate?

will provide more comments + IR test

kbobrovs commented on sycl/include/CL/sycl/detail/util.hpp:
```diff
@@ -30,6 +31,34 @@ class Sync {
+// Interface to iterate via C strings.
+class CStringIterator {
+public:
+  // Get the next string. Retuns next string's pointer or nullptr.
```
ok

kbobrovs commented on sycl/source/detail/program_impl.hpp:
```diff
@@ -293,6 +293,15 @@ class program_impl {
+                              size_t ValSize);
+
+  /// Returns the OS module handle this program belogs to. A program belongs to
+  /// an OS module if it was built from device image(s) beloning to that module.
```
ok

kbobrovs commented on sycl/source/detail/spec_constant_impl.cpp:
```diff
@@ -0,0 +1,42 @@
+    Out << (I == 0 ? "" : " ") << static_cast<int>(*(V.getValuePtr() + I));
+  }
+  Out << "]";
+  Out << std::dec;
```
ok

kbobrovs commented on sycl/test/aot/spec_const_aot.cpp:
```diff
@@ -0,0 +1,73 @@
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
```
nevertheless, all (most) other tests have it; let's leave it like that for now, then adjust if needed in a bulk operation over all tests

kbobrovs commented:
> @kbobrovs kbobrovs force-pushed the kbobrovs:spec-const-ALL branch from 2223f8c to dc40485 2 minutes ago:

- Applied clang format
- Added spec_const_redefine.cpp test (under XFAIL for now)
- Re-enabled spec_const_hw.cpp on CPU device (required llvm-spirv fixes are now in place)

kbobrovs commented:
Dear reviewers - @AGindinson, @erichkeane, @AlexeySachkov, @Fznamznon, @sndmitriev, @smaslov-intel, @romanovvlad, @mdtoguchi, @sergey-semenov - please take another look. All prior comments have been addressed to the best of my knowledge.

kbobrovs commented:
> @kbobrovs kbobrovs force-pushed the kbobrovs:spec-const-ALL branch from 4e0b174 to 316a855 38 minutes ago

rebase, squashed deploy dependence commit, replaced first three commits with updated after code review. 2 review comments from @erichkeane to be fixed in FE

kbobrovs commented:
> kbobrovs force-pushed the kbobrovs:spec-const-ALL branch from 316a855 to 8766334 3 hours ago

two review comments in FE (SemaSYCL.cpp) addressed

Fznamznon commented on llvm/tools/sycl-post-link/SpecConstants.cpp:
```diff
@@ -0,0 +1,296 @@
+static void AssertRelease(bool Cond, const char *Msg) {
+  if (!Cond)
+    llvm_unreachable((Twine("SpecConstants.cpp: ") + Msg).str().c_str());
+}
```
I did not get what is the purpose of this function. `llvm_unreachable` and standard c asserts work in the same way, i.e. in debug mode of when it's "build with asserts" configuration. In other words, using `llvm_unreachable" won't make it check something in release mode unless asserts are enabled.
My point was that we are doing some assertions only under
`#ifndef NDEBUG` so they are not presented in the code for `release with asserts` configuration.

kbobrovs replied:
Right, remnant from experiment. Will change to using llvm_unreachable directly.

kbobrovs replied:
> Will change to using llvm_unreachable directly.

I need to take this back, see below.

> In other words, using llvm_unreachable won't make it check something in release mode unless asserts are enabled.

Actually, the experiment showed that -DCMAKE_BUILD_TYPE=Release does set NDEBUG, and the effect of LLVM unreachable on Linux is a signal caught by handler - i.e. a crash. So it does work as release assert. And I don't want to write `if (Cond) llvm_unreachable();` everywhere in the code.

kbobrovs replied:
BTW, this indeed diverges from what https://llvm.org/docs/CodingStandards.html says: 
> Neither assertions or llvm_unreachable will abort the program on a release build.'

Fznamznon replied:
> Actually, the experiment showed that -DCMAKE_BUILD_TYPE=Release does set NDEBUG, and the effect of LLVM unreachable on Linux is a signal caught by handler - i.e. a crash. So it does work as release assert.
 
This is really strange! Assertions & llvm_unreachable should be triggered in Release build only if `LLVM_ENABLE_ASSERTIONS` cmake variable is set to `ON`...

Fznamznon replied:
Even the doc for `llvm_unreachable` says that it gives errors only in debug mode, https://llvm.org/doxygen/Support_2ErrorHandling_8h.html#ace243f5c25697a1107cce46626b3dc94

kbobrovs replied:
So:
- description in the docs does not match the  behavior
- behavior is "release assert"
- what actually was under question is just the name of the function, as LLVM anyway recommends unreachable over assert, and if it did not give crash in release, it would have been called CheckAndAssert or somthing.

Given that I believe the code should be left as is here

kbobrovs replied:
Please see below - replaced with report_fatal_error

erichkeane commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -1127,6 +1145,21 @@ static void populateIntHeader(SYCLIntegrationHeader &H, const StringRef Name,
+          cast<ClassTemplateSpecializationDecl>(ArgTy->getAsCXXRecordDecl());
+      const TemplateArgumentList *TemplateArgs =
+          &TmplSpec->getTemplateInstantiationArgs();
+      constexpr unsigned int TypeIDTmplArgNo = 1;
```
Whats the purpose of this?  Is it really only used the 1x below?  I think I'd rather it just be get(1)

erichkeane replied:
Also, do we just ignore the first argument?

kbobrovs replied:
> Whats the purpose of this? Is it really only used the 1x below? I think I'd rather it just be get(1)

It is self-commenting var name for the index '1'. Yes, it is used once, but my idea is that it is
better use properly named constant rather then literal, similarly to `#define SOME_DESCRIPTIVE_NAME val`

> Also, do we just ignore the first argument?

Here in FE - yes. It is used in the runtime though.

@erichkeane - given the explanations, are you OK with not changing the code here?

erichkeane replied:
IMO it doesn't make it more readable. I'd still prefer it just be get(1). Add a comment that it is the typeid if you want (actually a good descriptive comment would be appreciated).

kbobrovs replied:
ok

mdtoguchi commented on clang/include/clang/Driver/Action.h:
```diff
@@ -705,6 +714,47 @@ class BackendCompileJobAction : public JobAction {
 
+// Represents a file table transformation action. The order of inputs to a
+// FileTableTformJobAction was constructed with must accord with the tforms
+// added later - some tforms "consume" inputs. For example, "replace column"
```
I'm not parsing this comment - can you re-word?  `was constructed with` is not reading well.

kbobrovs replied:
ok

mdtoguchi commented on clang/lib/Driver/Action.cpp:
```diff
@@ -8,6 +8,7 @@
 
 #include "clang/Driver/Action.h"
 #include "llvm/Support/ErrorHandling.h"
+#include "llvm/Support/raw_ostream.h"
```
Is this needed?

kbobrovs replied:
looks like those two fixes got forgotten - will re-implement.

mdtoguchi commented on clang/lib/Driver/Driver.cpp:
```diff
@@ -3903,8 +3990,11 @@ class OffloadingActionBuilder final {
+                      DeviceCodeSplitArg->getValue() == StringRef("off");
+      DeviceCodeSplit =
+          !SplitOff && (DeviceCodeSplitArg ||
+                        Args.getLastArg(options::OPT_fsycl_device_code_split));
```
`OPT_fsycl_device_code_split` is an alias to `OPT_fsycl_device_code_split_EQ` so you only have to check if `OPT_fsycl_device_code_split_EQ` arg is set && !off (similar to what you did below)

kbobrovs replied:
ok

kbobrovs commented:
> @kbobrovs kbobrovs force-pushed the kbobrovs:spec-const-ALL branch from 49fa9bc to 3324fb3 6 minutes ago

addressed review comments from @erichkeane , @Fznamznon , @mdtoguchi

kbobrovs commented:
@bader, when merging, please merge these first:
    property I/O: #1357
    table library: #1358
    file-table-tform: #1359
then this one, excluding the first three:
@kbobrovs
Implement a property set I/O library. …
0f04616
@kbobrovs
Implement a simple string tabular data management library. …
36dff1c
@kbobrovs
Add a file-table-tform tool: manipulating tabular string data files. …
745170a

let me know if you need updates from me

Fznamznon commented on llvm/tools/sycl-post-link/sycl-post-link.cpp:
```diff
@@ -186,75 +222,95 @@ splitModule(Module &M,
+  StringRef Dir0 = OutputDir.getNumOccurrences() > 0
+                       ? OutputDir
+                       : sys::path::parent_path(OutputFilename);
+  const auto Sep = sys::path::get_separator();
```
Please do not use `auto` here.

kbobrovs replied:
ok

Fznamznon commented on llvm/tools/sycl-post-link/SpecConstants.cpp:
```diff
@@ -0,0 +1,296 @@
+    IRModified = IRModified || (SCIntrCalls.size() > 0);
+
+    for (auto *CI : SCIntrCalls) {
+      // 1. Find the ysmbolic ID (string literal) passed as the actual argument
```
```suggestion
      // 1. Find the symbolic ID (string literal) passed as the actual argument
```

kbobrovs replied:
ok

Fznamznon commented on llvm/tools/sycl-post-link/SpecConstants.cpp:
```diff
@@ -0,0 +1,296 @@
+  StringRef Res = cast<ConstantDataArray>(Init)->getAsString();
+  if (Res.size() > 0 && Res[Res.size() - 1] == '\0') {
+    Res = Res.substr(0, Res.size() - 1);
+  }
```
You can drop curly braces here.

kbobrovs replied:
ok

kbobrovs commented:
> Even the doc for llvm_unreachable says that it gives errors only in debug mode, https://llvm.org/doxygen/Support_2ErrorHandling_8h.html#ace243f5c25697a1107cce46626b3dc94

see my reply
https://github.com/intel/llvm/pull/1356#discussion_r397988458

kbobrovs commented:
> > Even the doc for llvm_unreachable says that it gives errors only in debug mode, https://llvm.org/doxygen/Support_2ErrorHandling_8h.html#ace243f5c25697a1107cce46626b3dc94
> 
> see my reply
> [#1356 (comment)](https://github.com/intel/llvm/pull/1356#discussion_r397988458)

added a TODO in the code

Fznamznon commented:
BTW, It would be nice to have spec constants design explanation somewhere in this doc https://github.com/intel/llvm/blob/sycl/sycl/doc/CompilerAndRuntimeDesign.md someday.

kbobrovs commented:
@Fznamznon, @sndmitriev, @smaslov-intel, @romanovvlad, @mdtoguchi - please approve post-link, offload wrapper, PI, RT and driver changes if you are OK with current code

Fznamznon commented:
> @Fznamznon, @sndmitriev, @smaslov-intel, @romanovvlad, @mdtoguchi - please approve post-link, offload wrapper, PI, RT and driver changes if you are OK with current code

Post link part is approved https://github.com/intel/llvm/pull/1356#pullrequestreview-381781724

mdtoguchi commented:
LGTM for Driver.  Thanks!

kbobrovs commented:
> BTW, It would be nice to have spec constants design explanation somewhere in this doc https://github.com/intel/llvm/blob/sycl/sycl/doc/CompilerAndRuntimeDesign.md someday.

yes, agree. But let's do this after the API and the implementation is finalized.

kbobrovs commented:
> @kbobrovs kbobrovs force-pushed the kbobrovs:spec-const-ALL branch from 5ef5cae to eb382f7 5 minutes ago

resolved a conflict in comment due to change in #1359

romanovvlad commented on sycl/CMakeLists.txt:
```diff
@@ -210,7 +210,7 @@ add_custom_target( sycl-toolchain
           clang-offload-wrapper
           clang-offload-bundler
-	  file-table-tform
+          file-table-tform
```
test

kbobrovs replied:
??

romanovvlad commented on sycl/source/detail/program_manager/program_manager.cpp:
```diff
@@ -590,54 +598,81 @@ ProgramManager::ProgramManager() {
 
-    if (DbgProgMgr > 0)
-      std::cerr << "loaded device image from " << SpvFile << "\n";
+    if (Imgs.size() == 0)
```
Minor. How can it happen that Imgs.size() is zero here?

kbobrovs replied:
Looks like it can't happen - will change to assert.

kbobrovs commented:
@bader, I've got approvals from component owners - please take a look and merge.
